cli/flox-rust-sdk/src/models/manifest/composite.rs

```diff
@@ -0,0 +1,280 @@
+#![allow(dead_code)] // TODO: Remove on first use.
+
+use std::iter::once;
+
+use flox_core::Version;
+use thiserror::Error;
+
+use super::typed::{
+    Manifest,
+    ManifestBuild,
+    ManifestContainerize,
+    ManifestHook,
+    ManifestInstall,
+    ManifestOptions,
+    ManifestProfile,
+    ManifestServices,
+    ManifestVariables,
+};
+
+#[derive(Error, Debug)]
+pub enum MergeError {}
+
+/// A collection of manifests to be merged with a `ManifestMergeStrategy`.
+struct CompositeManifest {
+    composer: Manifest,
+    deps: Vec<Manifest>,
+}
+
+impl CompositeManifest {
+    fn merge_all(&self, merger: impl ManifestMergeStrategy) -> Result<Manifest, MergeError> {
+        let Some(first_dep) = self.deps.first() else {
+            // No deps, just composer.
+            return Ok(self.composer.clone());
+        };
+
+        self.deps
+            .iter()
+            .skip(1) // First dep is used as initializer.
+            .chain(once(&self.composer)) // Composer goes last.
+            .try_fold(first_dep.clone(), |merged, next| {
+                merger.merge(merged, next.clone())
+            })
+    }
+}
+
+/// Strategy for merging two manifests which can then be applied iteratively for
+/// multiple manifests.
+trait ManifestMergeStrategy {
+    fn merge_version(manifest1: &Manifest, manifest2: &Manifest) -> Result<Version<1>, MergeError>;
+    fn merge_install(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestInstall, MergeError>;
+    fn merge_vars(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestVariables, MergeError>;
+    fn merge_hook(manifest1: &Manifest, manifest2: &Manifest) -> Result<ManifestHook, MergeError>;
+    fn merge_profile(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestProfile, MergeError>;
+    fn merge_options(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestOptions, MergeError>;
+    fn merge_services(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestServices, MergeError>;
+    fn merge_build(manifest1: &Manifest, manifest2: &Manifest)
+        -> Result<ManifestBuild, MergeError>;
+    fn merge_containerize(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<Option<ManifestContainerize>, MergeError>;
```

Comment:
are these supposed to take manifests each instead of the actual sections they are merging?

Reply:
Done (squashed because it will be hard to rebase across commits)